tools/data_integrity_check.php

```diff
@@ -0,0 +1,129 @@
+$instrument_arr = $DB->pselectCol("SELECT Test_name FROM test_names", []);
+$check_arr      = [];
+
+$helper->printSuccess('Getting data from `test_names` table...');
+$instruments       = array_filter(
+    $instrument_arr,
+    [$DB, 'tableExists']
+);
+$numberOfTestNames = count($instrument_arr);
+$numberOfTestNamesWithTables = count($instruments);
+if ($numberOfTestNames !== $numberOfTestNamesWithTables) {
+    $helper->printError(
+        'Found entries in the `test_names` table that do not have appear to '
+        . 'be installed: ' . join(', ', array_diff($instrument_arr, $instruments))
+    );
+}
+
+
+$helper->printSuccess("Looking for orphaned instrument/flag");
+foreach ($instruments as $instrument) {
+    $helper->printLine("Checking instrument `$instrument`...");
+
+    $orphan_flag_arr = $DB->pselect(
+        "
+        SELECT
+            flag.*
+        FROM
+            flag
+        LEFT JOIN
+            {$instrument}
+        ON
+            {$instrument}.CommentID = flag.CommentID
+        WHERE
+            Test_name = :instrument AND
+            {$instrument}.CommentID IS NULL
+    ",
+        array(
+            "instrument" => $instrument
+        )
+    );
+
+    $flag_str = $instrument . " flag";
+    foreach ($orphan_flag_arr as $orphan_flag) {
+        $comment_id = $orphan_flag["CommentID"];
+        $helper->printWarning("orphan flag {$flag_str}:{$comment_id}");
+    }
+    $orphan_instrument_arr = $DB->pselect(
+        "
+        SELECT
+            {$instrument}.*
+        FROM
+            {$instrument}
+        LEFT JOIN
+            flag
+        ON
+            {$instrument}.CommentID = flag.CommentID
+        WHERE
+            flag.CommentID IS NULL
+    ",
+        array()
+    );
+
+    foreach ($orphan_instrument_arr as $orphan_instrument) {
+        $comment_id = $orphan_instrument["CommentID"];
+        $helper->printWarning("orphan instrument {$instrument}:{$comment_id}");
+    }
+}
+
+$helper->printSuccess("Checking for duplicate flags in a session");
+$duplicate_flag_arr = $DB->pselect(
+    "
+    SELECT
+        SessionID,Test_name,CommentID
+    FROM
+        flag
+    WHERE
+        (
+            flag.CommentID NOT LIKE 'DDE_%' AND
+            (
+                SELECT
+                    COUNT(*)
+                FROM
+                    flag test
+                WHERE
+                    test.CommentID NOT LIKE 'DDE_%' AND
+                    flag.SessionID = test.SessionID AND
+                    flag.Test_name = test.Test_name
+            ) > 1
+        ) OR
+        (
+            flag.CommentID LIKE 'DDE_%' AND
+            (
+                SELECT
+                    COUNT(*)
+                FROM
+                    flag test
+                WHERE
+                    test.CommentID LIKE 'DDE_%' AND
+                    flag.SessionID = test.SessionID AND
+                    flag.Test_name = test.Test_name
+            ) > 1
+        )
```

Comment:
Wouldnt a query grouping by session and test_name with a count do the same thing here ?

it would be significantly quicker i think

Reply:
could use `having count(*) > 1` with the `group by` clause

Reply:
Could one of you provide more details? SQL isn't my strong suit

Reply:
@johnsaigle Can do but would like to know which table is that "test" table in the cross-join? Cross-join should almost never be used except some specific case.

Reply:
I think in this statement `test` is an alias of the `flag` table. I adapted this query from the original PR (linked in description) which I did not write.